src/gui/plugins/apply_force_torque/ApplyForceTorque.cc

```diff
@@ -0,0 +1,336 @@
+    public: bool changedEntity{false};
+
+    /// \brief Force to be applied
+    public: math::Vector3d force{0.0, 0.0, 0.0};
+
+    /// \brief Offset of the force application relative to center of mass
+    public: math::Vector3d offset{0.0, 0.0, 0.0};
+
+    /// \brief Torque to be applied
+    public: math::Vector3d torque{0.0, 0.0, 0.0};
+  };
+}
+}
+
+using namespace gz;
+using namespace sim;
+
+/////////////////////////////////////////////////
+ApplyForceTorque::ApplyForceTorque()
+  : GuiSystem(), dataPtr(std::make_unique<ApplyForceTorquePrivate>())
+{
+}
+
+/////////////////////////////////////////////////
+ApplyForceTorque::~ApplyForceTorque() = default;
+
+/////////////////////////////////////////////////
+void ApplyForceTorque::LoadConfig(const tinyxml2::XMLElement */*_pluginElem*/)
+{
+  if (this->title.empty())
+    this->title = "Apply force and torque";
+
+  gz::gui::App()->findChild<gz::gui::MainWindow *>
+      ()->installEventFilter(this);
+  gz::gui::App()->findChild<gz::gui::MainWindow *>
+      ()->QuickWindow()->installEventFilter(this);
+}
+
+/////////////////////////////////////////////////
+bool ApplyForceTorque::eventFilter(QObject *_obj, QEvent *_event)
+{
+  if (_event->type() ==
+    gz::sim::gui::events::EntitiesSelected::kType)
+  {
+    gz::sim::gui::events::EntitiesSelected *_e =
+        static_cast<gz::sim::gui::events::EntitiesSelected*>(_event);
+    this->dataPtr->selectedEntities = _e->Data();
+    this->dataPtr->changedEntity = true;
+  }
+  else if (_event->type() ==
+    gz::sim::gui::events::DeselectAllEntities::kType)
+  {
+    this->dataPtr->selectedEntities.clear();
+    this->dataPtr->changedEntity = true;
+  }
+
+  return QObject::eventFilter(_obj, _event);
+}
+
+/////////////////////////////////////////////////
+void ApplyForceTorque::Update(const UpdateInfo &/*_info*/,
+    EntityComponentManager &_ecm)
+{
+  {
+    std::lock_guard<std::mutex> lock(this->dataPtr->mutex);
+
+    if (!this->dataPtr->changedEntity)
+    {
+      return;
+    }
+    this->dataPtr->changedEntity = false;
+
+    this->dataPtr->modelName = "";
+    this->dataPtr->linkList.clear();
+    this->dataPtr->linkNameList.clear();
+    this->dataPtr->linkIndex = -1;
+
+    if (!this->dataPtr->selectedEntities.empty())
+    {
+      auto entity = this->dataPtr->selectedEntities.front();
+      Model selectedModel(entity);
+      Link selectedLink(entity);
+      if (selectedModel.Valid(_ecm))
+      {
+        selectedLink = Link(selectedModel.CanonicalLink(_ecm));
+      }
+      else if (selectedLink.Valid(_ecm))
+      {
+        selectedModel = *selectedLink.ParentModel(_ecm);
+      }
+      else
+      {
+        return;
+      }
+
+      this->dataPtr->modelName = QString::fromStdString(
+        selectedModel.Name(_ecm));
+
+      // Put all of the model's links into the list
+      auto links = selectedModel.Links(_ecm);
+      unsigned int i{0};
+      while (i < selectedModel.LinkCount(_ecm))
```

Comment:
nitpick:
```suggestion
      while (i < links.size())
```
since we're actually indexing into `links`.